NET-423: Update client README

bearni95 opened the pull request:
See proposed changes in the [linear ticket](https://linear.app/streamr/issue/NET-423/update-client-documentation-including-5x-6x-changelog)

bearni95 committed:
docs: 'usage' section revised

bearni95 committed:
docs: renamed client options into client configuration; indexed all the config options currently available for the client

bearni95 committed:
docs: 'authentication'; 'managing subscriptions'

bearni95 committed:
docs: stream subscriptions, storage

bearni95 committed:
docs: storage, stream API

bearni95 committed:
docs: events (should these be gone?); partitioning, almost untouched

bearni95 committed:
docs: updated events warning

bearni95 committed:
docs: updates, formating;

bearni95 committed:
docs: modifications to 'Client configuration'; minor tweaks

bearni95 committed:
docs: formatting

bearni95 committed:
docs: changed subsection into 'Get started'; break down of browser links for corea vs brubeck; permissions/operations table; tested utility fns, updated return types to Promise

bearni95 committed:
updated docs to 6.0.0-beta.2

bearni95 committed:
docs: fixed typo

harbu commented on packages/client/README.md:
````diff
@@ -19,51 +19,100 @@ Please see the [Streamr project docs](https://streamr.network/docs) for more det
+```typescript
+import { StorageNode, StreamrClient } from StreamrClient
+
+await stream.addToStorageNode(StorageNode.STREAMR_GERMANY)
````
Update to reference `STREAMR_STORAGE_NODE_GERMANY`

harbu commented on packages/client/README.md:
```diff
@@ -19,51 +19,100 @@ Please see the [Streamr project docs](https://streamr.network/docs) for more det
-    stream: 'streamId',
-    partition: 0, // Optional, defaults to zero. Use for partitioned streams to select partition.
+    stream: STREAM_ID,
+    streamPartition: 0, // Optional, defaults to zero. Use for partitioned streams to select partition.
```
```suggestion
    id: STREAM_ID,
    partition: 0, // Optional, defaults to zero. Use for partitioned streams to select partition.
```

harbu commented on packages/client/README.md:
````diff
@@ -73,10 +122,10 @@ const sub = await client.subscribe({
+```typescript
+const sub = await client.resendSubscribe({
+    stream: STREAM_ID,
+    resend: {
````
This should be changed to be 
```ts
client.resend(
    'streamId',
    {
        last : 5
    },
    (message) => {
        .... // whatever is here
    }
)
```
in accordance with aa7b7a0c3db3cf972355629390f3798dcb2d0c4e.

harbu commented on packages/client/README.md:
````diff
@@ -19,51 +19,100 @@ Please see the [Streamr project docs](https://streamr.network/docs) for more det
+
+<!-- for Brubeck package (6x) -->
+<script src="https://unpkg.com/streamr-client@6.0.0-beta.2/streamr-client.web.js"></script>
+```
````
```suggestion
For usage in the browser include the latest build, e.g. by including a <script> tag pointing at a CDN:
```html
<!-- for Brubeck package (6x) -->
<script src="https://unpkg.com/streamr-client@6.0.0-beta.2/streamr-client.web.js"></script>
```
```

harbu commented on packages/client/README.md:
````diff
@@ -19,51 +19,100 @@ Please see the [Streamr project docs](https://streamr.network/docs) for more det
+const stream = await client.createStream({
+    id: STREAM_ID
+})
+```
````
Maybe worth outlining the two ways of creating a stream:
```suggestion
const stream = await client.createStream({
    id: '/foo/bar'
})
// stream.id === `${address}/foo/bar`
```

teogeb replied:
Agree with @harbu on this

harbu commented on packages/client/README.md:
```diff
@@ -19,51 +19,100 @@ Please see the [Streamr project docs](https://streamr.network/docs) for more det
-## Installation
+>In Streamr, Ethereum accounts are used for identity. You can generate an Ethereum private key using any Ethereum wallet, or you can use the utility function `StreamrClient.generateEthereumAccount()`, which returns the address and private key of a fresh Ethereum account.
 
+### Installation
```
Before section `Installation` two quick snippets:
- subscribing to a stream
- creating & publishing to a stream

teogeb replied:
And the snippets should be in order of importance (subscribe is more typical), as it is already in Usage section

harbu commented on packages/client/README.md:
```diff
@@ -114,49 +148,226 @@ const msg = {
 ----
 
-## Client options
+## Client configuration
```
- Let's get rid of the table, we shall replace it with TSdocs in the future.
- Let's get rid of JSON as well.
- Just explain `auth` and `orderMessages` and `gapFill`.

harbu commented on packages/client/README.md:
```diff
@@ -114,49 +148,226 @@ const msg = {
-## Authentication options
+## Authentication
 
 Note: **Authenticating with an API key has been deprecated. Cryptographic keys/wallets is the only supported authentication method.**
```
Just drop note

harbu commented on packages/client/README.md:
```diff
@@ -19,51 +19,100 @@ Please see the [Streamr project docs](https://streamr.network/docs) for more det
 [Installation](#installation) · [Usage](#usage) · [API Docs](#API-docs) · [Client options](#client-options) · [Authentication](#authentication-options) · [Managing subscriptions](#managing-subscriptions) · [Stream API](#stream-api) · [Subscription options](#subscription-options) · [Storage](#storage) ·[Data Unions](#data-unions) · [Utility functions](#utility-functions) · [Events](#events) · [Str ...
 
+## Get Started
+Here are some usage examples. More examples can be found [here](https://github.com/streamr-dev/examples).
```
Somewhere in the beginning of this document, advertise https://github.com/streamr-dev/streamr-client-react for react use cases.

harbu replied:
Let's keep this in hold for now, we should understand when the react thing gets released.

harbu commented on packages/client/README.md:
````diff
@@ -166,17 +377,28 @@ const client = new StreamrClient({
+})
+```
+
 Authenticating with a pre-existing session token (used internally by the Streamr app):
````
You could shortly explain that a `StreamrClient` can be created without a private key (needs testing that this actually works):
```javascript
const client = new StreamrClient()
```
It will be associated with a randomly generated address (that does not get stored anywhere, a bit of a throw away account), and that has limited capability due to not having gas.

harbu commented on packages/client/README.md:
````diff
@@ -166,17 +377,28 @@ const client = new StreamrClient({
 Authenticating with a pre-existing session token (used internally by the Streamr app):
 
-```js
+```typescript
````
Just remove all these stuff about `sessionToken`. (It is technically still in the code, but something we don't advertise.)

harbu commented on packages/client/README.md:
```diff
@@ -231,7 +453,7 @@ const client = new StreamrClient({
-## Managing subscriptions
+## Stream subscriptions
 
 | Name                         | Description                                                                                                                                                                     |
```
Get rid of table, add code snippet for cases that aren't already included below.

harbu commented on packages/client/README.md:
```diff
@@ -240,6 +462,23 @@ await client.connect()
+const subscription = await client.subscribe(
+    { stream: STREAM_ID },
+    (message) => { ... }
+)
```
```suggestion
const subscription = await client.subscribe(STREAM_ID, (message) => { ... })
```

harbu commented on packages/client/README.md:
```diff
@@ -240,6 +462,23 @@ await client.connect()
+const subscriptions = client.getSubscriptions()
+
+// unsubscribing from an existent subscription:
+await client.unsubscribe({stream: STREAM_ID})
```
```suggestion
await client.unsubscribe(STREAM_ID)
```

harbu commented on packages/client/README.md:
```diff
@@ -240,6 +462,23 @@ await client.connect()
+await client.unsubscribe({stream: STREAM_ID})
+
+// or, unsubscribe them all:
+const streams = await client.removeAll()
```
To unsubscribe from all streams
```suggestion
const streams = await client.unsubscribe()
```

harbu commented on packages/client/README.md:
```diff
@@ -249,10 +488,10 @@ The second argument to `client.subscribe(options, callback)` is the callback fun
 | payload       | A JS object containing the message payload itself                                                                                                                                                                                |
```
Remove table, and move to snippet

harbu commented on packages/client/README.md:
```diff
@@ -249,10 +488,10 @@ The second argument to `client.subscribe(options, callback)` is the callback fun
-    streamId: 'my-stream-id',
-}, (payload, streamMessage) => {
+    streamId: STREAM_ID,
+}, (payload: any, streamMessage: StreamMessage) => {
```
```suggestion
}, (payload, streamMessage) => { // streamMessage is of type StreamMessage
```

bearni95 committed:
Merge branch 'main' into NET-423

harbu commented on packages/client/README.md:
```diff
@@ -271,18 +510,18 @@ Note that only one of the resend options can be used for a particular subscripti
 | resend    | Object defining the resend options. Below are examples of its contents.            |
```
Get rid of table.

harbu commented on packages/client/README.md:
```diff
@@ -271,18 +510,18 @@ Note that only one of the resend options can be used for a particular subscripti
 // Resend N most recent messages
-const sub1 = await client.subscribe({
-    streamId: 'my-stream-id',
+const sub1 = await client.subscribeResend({
```
```suggestion
const sub1 = await client.subscribe({
```

harbu commented on packages/client/README.md:
```diff
@@ -271,18 +510,18 @@ Note that only one of the resend options can be used for a particular subscripti
 // Resend from a specific message reference up to the newest message
-const sub2 = await client.subscribe({
-    streamId: 'my-stream-id',
+const sub2 = await client.subscribeResend({
```
```suggestion
const sub2 = await client.subscribe({
```

harbu commented on packages/client/README.md:
```diff
@@ -271,18 +510,18 @@ Note that only one of the resend options can be used for a particular subscripti
+    streamId: STREAM_ID,
     resend: {
         from: {
             timestamp: 12345,
```
Remove `msgChainId` here.

harbu commented on packages/client/README.md:
```diff
@@ -249,10 +488,10 @@ The second argument to `client.subscribe(options, callback)` is the callback fun
+}, (payload: any, streamMessage: StreamMessage) => {
     console.log({
         payload, streamMessage
     })
```
Change title below
> ### Subscription Options
to something like `Doing a resend with subscription`

harbu commented on packages/client/README.md:
```diff
@@ -310,21 +549,35 @@ const sub3 = await client.subscribe({
+// COULD NOT FIND A REPLACEMENT
 const sub = await client.subscribe(options)
 sub.on('resent', () => {
     console.log('All caught up and received all requested historical messages! Now switching to real time!')
```
```suggestion
sub.onResent(() => {
    console.log('All caught up and received all requested historical messages! Now switching to real time!')
})
```

harbu commented on packages/client/README.md:
````diff
@@ -310,21 +549,35 @@ const sub3 = await client.subscribe({
-```js
+```typescript
+// THIS CODE DOES NOT CURRENTLY WORK! 
+// COULD NOT FIND A REPLACEMENT
````
Can remove this now after the change below

harbu commented on packages/client/README.md:
```diff
@@ -310,21 +549,35 @@ const sub3 = await client.subscribe({
+import { StorageNode } from 'streamr-client'
+...
+// assign a stream to storage
+await stream.addToStorageNode(StorageNode.STREAMR_GERMANY)
```
Change constant

harbu commented on packages/client/README.md:
```diff
@@ -310,21 +549,35 @@ const sub3 = await client.subscribe({
+// fetch the storage nodes for a stream
+const storageNodes = stream.getStorageNodes()
+// remove the stream from a storage node
+await stream.removeFromStorageNode(StorageNode.STREAMR_GERMANY)
```
Change constant

harbu commented on packages/client/README.md:
```diff
@@ -333,37 +586,93 @@ All the below functions return a Promise which gets resolved with the result.
+| listStreams(query)                                  | Fetches an array of stream objects from the API. For the query params, consult the [API docs](https://api-explorer.streamr.com).                     |
+| getStreamByName(name)                               | Fetches a stream which exactly matches the given name.                                                                                               |
+| createStream(\[properties])                         | Creates a stream with the given properties. For more information on the stream properties, consult the [API docs](https://api-explorer.streamr.com). You must specify the `id`. It must start with your ethereum address.|
+| getOrCreateStream(properties)                       | Gets a stream with the id or name given in `properties`, or creates it if one is not found.                                                          |
```
The left-side (the removed stuff here) seems like the correct versions. However, we probably should remove the table anyways.

harbu commented on packages/client/README.md:
```diff
@@ -333,37 +586,93 @@ All the below functions return a Promise which gets resolved with the result.
+const stream: Stream = await client.getStream(STREAM_ID)
+
+// getStreamByName
+const stream = await client.getStreamByName('stream-name')
```
This no longer exists. Remove.

harbu commented on packages/client/README.md:
```diff
@@ -333,37 +586,93 @@ All the below functions return a Promise which gets resolved with the result.
+    grantedAccess?: boolean;
+    publicAccess?: boolean;
+    operation?: StreamOperation;
+})
```
It's called `searchStreams` now and supports a different set of options.
```suggestion
const streams: Stream[] = await client.searchStreams({
    term: '/foo',
    permissionFilter: {
        allowPublic: true
    }
})
```

harbu commented on packages/client/README.md:
```diff
@@ -333,37 +586,93 @@ All the below functions return a Promise which gets resolved with the result.
+const stream = await client.getOrCreateStream({
+    id: `${await client.getAddress()}/stream-01`
+})
+
```
Already covered earlier, remove.

harbu commented on packages/client/README.md:
```diff
@@ -333,37 +586,93 @@ All the below functions return a Promise which gets resolved with the result.
+
+
+
 | Name                                      | Description                                                                                                                                                                                                                                                                   |
```
Replace table with code snippets if there are no code snippets for these methods. Ignore permission-related methods.

teogeb replied:
Permission snippets are here: https://github.com/streamr-dev/network-monorepo/pull/450/files

If that PR gets landed to `main` before this PR, please merge those snippets to this PR.

harbu commented on packages/client/README.md:
```diff
@@ -555,15 +864,17 @@ a Data Union deployed by a particular address with particular "name" will have a
 Note that this private key can be used to authenticate to the Streamr API
 by passing it in the authentication options, as described earlier in this document.
 
 ## Events
```
I'd remove this section, internal stuff.

harbu commented on packages/client/README.md:
````diff
@@ -607,9 +918,9 @@ Partitioning (sharding) enables streams to scale horizontally. This section desc
+```typescript
 const stream = await client.createStream({
-    name: 'My partitioned stream',
+    id: `${await client.getAddress()}/partitioned-stream`,
````
```suggestion
    id: `/partitioned-stream`,
```

harbu commented on packages/client/README.md:
````diff
@@ -623,8 +934,8 @@ The library allows the user to choose a _partition key_, which simplifies publis
-```js
-await client.publish('my-stream-id', msg, Date.now(), msg.vehicleId)
+```typescript
+await client.publish(STREAM_ID, msg, Date.now(), msg.vehicleId)
````
I would add to the above description an explicit note that `msg.vehicleId` is the partition key in this example.
```suggestion
await client.publish(STREAM_ID, msg, Date.now(), msg.vehicleId) // msg.vehicleId is the partition key here
```

teogeb replied:
Agree with @harbu here

harbu commented on packages/client/README.md:
````diff
@@ -634,9 +945,9 @@ await stream.publish(msg, Date.now(), msg.vehicleId)
+```typescript
 const sub = await client.subscribe({
-    stream: 'my-stream-id',
+    stream: STREAM_ID,
````
```suggestion
    id: STREAM_ID,
```

harbu commented on packages/client/README.md:
```diff
@@ -634,9 +945,9 @@ await stream.publish(msg, Date.now(), msg.vehicleId)
 const sub = await client.subscribe({
-    stream: 'my-stream-id',
+    stream: STREAM_ID,
     partition: 4, // defaults to 0
```
```suggestion
    partition: 4
```

harbu commented on packages/client/README.md:
```diff
@@ -634,9 +945,9 @@ await stream.publish(msg, Date.now(), msg.vehicleId)
 
 By default, the JS client subscribes to the first partition (partition `0`) in a stream. The partition number can be explicitly given in the subscribe call:
```
This behavior will change in the future so that it will subscribe to all partitions by default.

harbu commented on packages/client/README.md:
```diff
@@ -645,14 +956,14 @@ const sub = await client.subscribe({
 await Promise.all([2, 3, 4].map(async (partition) => {
     await client.subscribe({
-        stream: 'my-stream-id',
+        stream: STREAM_ID,
```
```suggestion
        id: STREAM_ID,
```

bearni95 committed:
docs: updated reference to storage node

bearni95 committed:
docs: explained vehicleId as partition key; replaced ts snippets with js

bearni95 committed:
docs: added reference to streamr-client-react; moved auth down one level, simplified snippets; removed refs to sessionToken; updated stream subscriptions snippet; updated streamid pattern;

bearni95 committed:
docs: removed api key note under authentication; added local url for StreamrClient.generateEthereumAccount

bearni95 committed:
docs: removed table and upgraded snippets for stream subscriptions

bearni95 committed:
docs: removed msg handler callback's table; upgraded the snippet and explanations accordingly

bearni95 committed:
docs: polished subcribe snippet for msg handler callback

bearni95 committed:
docs: reworked resend section

bearni95 committed:
docs: changed ts imports into js requires

bearni95 committed:
docs: changed more refs to streamr_storage_node_germany

bearni95 committed:
docs: client API snippet

bearni95 committed:
docs: stream API table moved into snippets

bearni95 committed:
docs: added 'gas required' comment on snippets that do so

bearni95 committed:
docs: redone snippets for stream object

bearni95 committed:
docs: removed utility functions for token balances; formated comment on generateEthereumAccount

bearni95 committed:
docs: subscription to partitioned streams 'subscribe' call reformated

bearni95 committed:
docs: upgraded explanation for partitioned streams def behaviour

bearni95 committed:
docs: added notice to remove 'important information' before launch; expanded info on streamId type

bearni95 committed:
Merge branch 'main' into NET-423

bearni95 committed:
docs: integrated latest permissions api

bearni95 committed:
docs: gapFill and ordering updates

bearni95 committed:
docs: added auth options for anon

bearni95 committed:
docs: cleanup for permissions

bearni95 committed:
docs: moved connecting into a subsection, removed table in favor of extended snippets

bearni95 committed:
docs: formatting fix

bearni95 committed:
docs: minor formatting fixes; reformated utility fns into snippets

bearni95 committed:
docs: comment order change

bearni95 committed:
docs: updated snippets from working examples

bearni95 committed:
docs: formatting

harbu commented on packages/client/README.md:
````diff
@@ -18,51 +18,133 @@ This library allows you to easily interact with the [Streamr Network](https://st
+    id: `${address}/foo/bar`,
+    partition: 1
+}
+```
````
*Forget this point, see teo's below*

As discussed in meeting, let's avoid talking about type `StreamDefinition` directly as it is a technicality from the end user's perspective. Instead something like "The public methods of the client generally support the following three ways of defining a stream:"
- Stream id `${address}/foo/bar`
- Stream id + partition `${address}/foo/bar#4`
- Stream id + partition `{ id:  `${address}/foo/bar`, partition: 4 }`

When defining a stream id only, the value of partition will be interpreted to take on a default value which is dependent on the method being called.

teogeb replied:
We could do this explaining in partitioning section. And all examples before the partitionin sections would just use the plain string as a streamId (therefore defaulting to 0 partition).

bearni95 committed:
Merge branch 'main' into NET-423

bearni95 committed:
docs: corrections on content

- in import script for browser update from beta2 to beta3
- in browser check if beta tag is enough
- replace foo/bar in await stream.publish({foo: 'bar'})
- replace [GAS REQUIRED] by a more friendly and less agressive observation
- on createStream later comment for id, make it a console.log and a comment with the expected value
- storage section and resends are apart, bring them together
- getOrCreateStream -- standarize id def
- move imports into [package], [statics]
- text change:
"Authenticating with Ethereum also automatically creates an associated Streamr user, even if it doesn't already exist. Under the hood, the client will cryptographically sign a challenge to authenticate you as a Streamr user:"
delete and explain like "you require a eth wallet, either pkey or window.ethereum, or just use the utility fn"
- remove from anon spinnet's text last portion " using an internally-generated disposable wallet:"
- message ordering: if your use-case doesn't require order of messages you can disable.
- in case of out-of-sync messaging your app will receive messages faster without these 2
- doesnt care about missing msgs or their order
- "tolerant" of missing messages as phrasing
- rework message ordering and gap filling as a sngle one, with orderMessages and gapFill set to false
- make "By default the client will automatically connect and disconnect as needed, ideally you should not need to manage connection state explicitly" into it's own paragraph
- explain the callback types under the above section (stream subs) and remove ""Message handler callback
- subscribe/resend explain the diff between .subscribe, .resend
- "by default subscribe will not resend historical data, only sub to real time messages..." as the intro text for "Resend functionality with subscriptions"
- resend snippets:
    - working with the timestamps to be "Date.now() - 1000 * 60 * 5" and have their
    - remove sequenceNumber
    - publisher should be a eth address
    - msgChainId get an example value from printing a sent streamr message.
    - publisher and msgChainId are required in tandem, but optional as a whole
- get rid of getStream example
- get rid of client.publish example
- add a searchStream example for plain term, allOf, anyOf
- add the user field
- in the "stream object" section get rid of getStream and publish
- remove comment with permissions = [{}, {}]
- put a coma!
- stream object: break down snippets into their own with an explanation/title
// And for public streams
await stream.grantPermissions({
    public: true
    permissions: [StreamPermission.SUBSCRIBE],
})
- stream operations and permissions: update "can..." to current naming
- remove stream ops and perms table and have a StreamPermission values breakdown as a snippet
- utility fns into 2 snippets, with explanation and title
-stream partitioning: 100 is the hard limit now
- remove "for devs" section
- on the Promise.all snippet rename "handler" into "messageCallback"
fixes and corrections:
- StreamPermission.canSubscribe --> .SUBSCRIBE
- partitioning: add example using the partition key and number
- partitioning 3 examples on publishing: using id, using partitionnumber as parameter, using partition key
- check that "In the browser, set localStorage.debug = 'StreamrClient*'" works
- (search queries) to build the use cases base it on the cli-tools: https://github.com/streamr-dev/network-monorepo/tree/main/packages/cli-tools
- rework teo's section "Stream permissions" into our existent permissions section

bearni95 committed:
docs: reformating the readme

bearni95 committed:
docs: reformat on readme proposal

bearni95 committed:
docs: formatting titles update

bearni95 committed:
docs: added note on stream id definition

bearni95 committed:
docs: correction on resend parameters

bearni95 committed:
docs: properly documented stream.update method

bearni95 committed:
docs: round of updates (see below)

- get rid of the "quick lookup section" entirely
- add the ts import along with "we spport it" underneath "import streamr-client"
- "When using Node.js remember to import the library with:" --> "To import using node.js..."
- have the react remember url be as normal text rather than after ">"
- move "use with react" section atop of the section
- "The Streamr Client API" into "Usage" (title)
- Move the stream object API as a section inside "usage"
- rename stream object api into "interacting with the `Stream` object" or something
- move data unions into the above api. utility functions as well
- "advanced usage" is a new section. stream partitioning, logging and proxying would go there.
- move the sections not covered by "usage" and that are under "quick lookup" inside their corresponding section
- rename "connecting" into "manual connection management"
- move "connecting" into advanced
- rename "message ordering" into "disable..." and move under advanced
- add benefits to disabling ordering "ur client won't wait for message filling/sorting, making your app quicker as it comes, rather than sorted. you may have missing messages"
- compare order of searchStreams snippets with the ones on cli-tools
- remove capitalized constants
- normalize address and stream variables

bearni95 committed:
docs: corrected id parameter; added example import for StreamPermission

fonty1 committed:
Update README.md

hanging bracket formatting

fonty1 committed:
Update README.md

fix brackets again :/

fonty1 committed:
Update README.md

- Tidying up getting started section. 
- "client" -> "streamr" (instance of a streamr-client)

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
 
 Please see the [Streamr project docs](https://streamr.network/docs) for more detailed documentation.
 
+## Contents
```
This could contain links to the relevant sections. Also we could include e.g. two levels of indentation to keep this simple. And e.g. Setup is a simple section, which doesn't need subsections in this TOC?

teogeb replied:
Links were added but reducing to max two level indentation not yet? And Setup subsections can be left out from this TOC.

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Subscribing to real-time events in a stream
+    - Publishing data points to a stream
+    - Resend functionality with subscriptions
+    - Search Streams
```
Consistency in casing for all titles?

harbu replied:
```suggestion
    - Searching for streams
```

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Creating a stream 
+    - Subscribing to real-time events in a stream
+    - Publishing data points to a stream
+    - Resend functionality with subscriptions
```
```suggestion
    - Requesting a resend of historical  events with subscriptions
```
This would highlight the historical nature of the data compared to the title "Subscribing to real-time events in a stream"

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Client Creation
+        - Authentication
+    - Creating a stream 
+    - Subscribing to real-time events in a stream
```
A subscription is always about real-time events, maybe no need to underline that aspect here?

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Resend functionality with subscriptions
+    - Search Streams
+    - Interacting with the `Stream` object
+        - Fetching existent streams
```
```suggestion
        - Fetching existing streams
```

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Installation
+    - Importing `streamr-client`
+- Usage
+    - Client Creation
```
```suggestion
    - Client creation
```

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Interacting with the `Stream` object
+        - Fetching existent streams
+        - Updating a stream
+        - Stream Permissions
```
```suggestion
        - Stream permissions
```

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+        - Authentication
+    - Creating a stream 
+    - Subscribing to real-time events in a stream
+    - Publishing data points to a stream
```
"Data points" is not relevant, let's keep this simple

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
 
+## Contents
+- Important information
+- Get Started
```
```suggestion
- Get started
```
consistent casing

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Search Streams
+    - Interacting with the `Stream` object
+        - Fetching existent streams
+        - Updating a stream
```
Update is less important than permissions

teogeb replied:
How was this resolved?

bearni95 replied:
Put permissions first, then update

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+        - Updating a stream
+        - Stream Permissions
+        - Deleting a stream
+    - Storage Options
```
```suggestion
    - Storing stream messages
```

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+        - Deleting a stream
+    - Storage Options
+    - Data Unions
+        - Admin Functions
```
```suggestion
        - Admin functions
```
consistent casing

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+        - Withdraw options
+        - Deployment options
+    - Utility functions
+- Advanced Usage
```
```suggestion
- Advanced usage
```

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+- Advanced Usage
+    - Manual connection management
+    - Disable message ordering
+    - Stream Partitioning
```
```suggestion
    - Stream partitioning
```

teogeb replied:
Stream partitionining is more important than ordering, which is more important than connection management?

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    - Manual connection management
+    - Disable message ordering
+    - Stream Partitioning
+        - A note on Stream ids and partitions
```
```suggestion
        - A note on stream ids and partitions
```

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+        - Updating a stream
+        - Stream Permissions
+        - Deleting a stream
+    - Storage Options
```
What does this mean? Should we indent this one step down

bearni95 replied:
Find better wording for "Storage Options" as a title

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+#### Authenticating
 
-## Installation
+In Streamr, Ethereum accounts are used for identity. You can generate an Ethereum private key using any Ethereum wallet, or you can use the utility function `StreamrClient.generateEthereumAccount()`, which returns the address and private key of a fresh Ethereum account. A private key is not required if you are only subscribing to public data on the Network. See more about [Authentication](#client ...
```
Should the link and the title match?

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+### Importing `streamr-client`
+To use with react please see [streamr-client-react](https://github.com/streamr-dev/streamr-client-react)
 
+If using typescript you can import the library with:
```
casing: TypeScript, consistent between "If"/"When"

harbu commented on packages/client/README.md:
````diff
@@ -10,97 +10,188 @@
+<!-- for Brubeck package (6x) -->
+<script src="https://unpkg.com/streamr-client@beta/streamr-client.web.js"></script>
+```
+___
````
Is this needed or some leftover thing?

teogeb commented on packages/client/README.md:
````diff
@@ -10,97 +10,188 @@
+For usage in the browser include the latest build, e.g. by including a `<script>` tag pointing at a CDN:
+
+```html
+<!-- for Brubeck package (6x) -->
````
Brubeck term is not needed, just leave out the comment

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+### Client Creation
+If you don't have an Ethereum account you can use the utility function [StreamrClient.generateEthereumAccount()](#utility-functions), which returns the address and private key of a fresh Ethereum account.
+
+More information on Stream IDs under the [stream creation project docs](https://streamr.network/docs/streams/creating-streams)
```
```suggestion
More information on Stream IDs can be found under the [stream creation project docs](https://streamr.network/docs/streams/creating-streams)
```

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+### Client Creation
+If you don't have an Ethereum account you can use the utility function [StreamrClient.generateEthereumAccount()](#utility-functions), which returns the address and private key of a fresh Ethereum account.
+
+More information on Stream IDs under the [stream creation project docs](https://streamr.network/docs/streams/creating-streams)
```
Maybe this could be stated a bit later?

teogeb replied:
Or as user maybe doesn't need authentication at all, we could leave out the paragraph here. If he wants a specific authentication, we surely has an Ethereum key already (and therefore doesn't need the utility function).

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+___
+## Usage
+### Client Creation
+If you don't have an Ethereum account you can use the utility function [StreamrClient.generateEthereumAccount()](#utility-functions), which returns the address and private key of a fresh Ethereum account.
```
I think rather than this paragraph, providing an example of using `StreamrClient.generateEthereumAccount()` in conjunction with creating a client could fit the theme of this section better?

harbu commented on packages/client/README.md:
````diff
@@ -10,97 +10,188 @@
 ```
 
-### Resending historical data
+Authenticating with an Ethereum private key contained in an Ethereum (web3) provider:
````
```suggestion
Authenticating with an Ethereum private key contained in an Ethereum (web3) provider (e.g. MetaMask):
```

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
-See "Subscription options" for resend options
-
-### Programmatically creating a stream
+You can also create an anonymous client instance that will be allowed to interact with public streams:
```
```suggestion
You can also create an anonymous client instance that can interact with public streams:
```

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
-// Optional: to enable historical data resends, add the stream to a storage node
-const { StorageNode, StreamrClient } = require('streamr-client').StreamrClient
-await stream.addToStorageNode(StorageNode.STREAMR_GERMANY)
+console.log(stream.id) // `${address}/foo/bar`
```
```suggestion
console.log(stream.id) // e.g. `0x1234567890123456789012345678901234567890/foo/bar`
```

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
 
-// Do something with the stream, for example call stream.publish(message)
+### Subscribing to real-time events in a stream
+The callback's first parameter, `payload`, will contain the value given to the `publish` method. The second parameter `streamrMessage` is of type StreamrObject. [You can read more about it here](../protocol/src/protocol/message_layer/StreamMessage.ts)
```
The sentence is referring to items that are introduced only later in the subsequent code snippet. For clarity, I would move the code snippet above this paragraph.

harbu replied:
```suggestion
The callback's first parameter, `payload`, will contain the contents of the message. The second parameter `streamrMessage` is of type `StreamMessage`, which contains additional metadata. [You can read more about `StreamMessage` here](../protocol/src/protocol/message_layer/StreamMessage.ts)
```

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
 
-// Do something with the stream, for example call stream.publish(message)
+### Subscribing to real-time events in a stream
+The callback's first parameter, `payload`, will contain the value given to the `publish` method. The second parameter `streamrMessage` is of type StreamrObject. [You can read more about it here](../protocol/src/protocol/message_layer/StreamMessage.ts)
```
Do we want to link to source code without explaining it more?

teogeb replied:
Ok, I suggest that we don't link to source code in this README. We can link to a API documentation when we have one. Also fix the StreamrObject -> `StreamrObject`.

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
 
-// Do something with the stream, for example call stream.publish(message)
+### Subscribing to real-time events in a stream
+The callback's first parameter, `payload`, will contain the value given to the `publish` method. The second parameter `streamrMessage` is of type StreamrObject. [You can read more about it here](../protocol/src/protocol/message_layer/StreamMessage.ts)
```
The first parameter is the content of the message, maybe rename the parameter to `content`?

teogeb replied:
We should use the same word `content` in all callbacks, see lines 755-764.

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
 
-// Do something with the stream, for example call stream.publish(message)
+### Subscribing to real-time events in a stream
+The callback's first parameter, `payload`, will contain the value given to the `publish` method. The second parameter `streamrMessage` is of type StreamrObject. [You can read more about it here](../protocol/src/protocol/message_layer/StreamMessage.ts)
```
Maybe the metadata can be explained in a separate paragraph, as it is a bit more advanced usage?

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+const subscription = await streamr.subscribe(
+    streamId, 
+    (payload, streamrMessage) => {
+        console.log(payload) // the value passed to the publish method
```
```suggestion
        console.log(payload) // the contents of the message
```

teogeb replied:
Remove both console.log comments, already explained

harbu commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+    streamId, 
+    (payload, streamrMessage) => {
+        console.log(payload) // the value passed to the publish method
+        console.log(streamrMessage) // the complete StreamrObject sent
```
```suggestion
        console.log(streamrMessage)
```

harbu commented on packages/client/README.md:
````diff
@@ -10,97 +10,188 @@
+Fetching all streams the client is subscribed to:
+```js
+const subscriptions = streamr.getSubscriptions()
+```
````
Unsubscribing is imo more relevant than fetching all streams the client is subscribed to, so I would move that example before this.

teogeb commented on packages/client/README.md:
````diff
@@ -10,97 +10,188 @@
+    }
+)
+```
+Fetching all streams the client is subscribed to:
````
Fetch would indicate a heavy operation

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
 
 // Publish with a specific timestamp as a Date object (default is now)
-await client.publish('my-stream-id', msg, new Date(54365472))
+await streamr.publish(streamId, msg, new Date(54365472))
```
Are these timestamps realistic data? Could use the timestamp of the other example

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+await streamr.publish(streamId, msg)
 
 // The first argument can also be the stream object
-await client.publish(stream, msg)
```
Could remove this use case. It is better to use `stream.publish()`.

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-
-Note: **Authenticating with an API key has been deprecated. Cryptographic keys/wallets is the only supported authentication method.**
+### Resend functionality with subscriptions
+By default `subscribe` will not resend historical data, only subscribe to real time messages. In order to fetch historical messages the stream needs to have [storage enabled](#storage).
```
```suggestion
By default `subscribe` will not request a resend of historical data, but only subscribe to real time messages. In order to fetch historical messages the stream needs to have [storage enabled](#storage).
```

harbu replied:
Actually I would move the 2nd sentence, "In order to fetch historical messages the stream needs to have storage enabled.", to occur after the `streamr.resend` and `streamr.subscribe` examples have been presented (before paragraph that starts with "Resend from a specific message reference up to the newest message:").

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+By default `subscribe` will not resend historical data, only subscribe to real time messages. In order to fetch historical messages the stream needs to have [storage enabled](#storage).
 
-If you don't have an Ethereum account you can use the utility function `StreamrClient.generateEthereumAccount()`, which returns the address and private key of a fresh Ethereum account.
+Note that only one of the resend options can be used for a particular subscription. The default functionality is to resend nothing, only subscribe to messages from the subscription moment onwards.
```
I would move this disclaimer after the different resend options have been introduced.

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+Note that only one of the resend options can be used for a particular subscription. The default functionality is to resend nothing, only subscribe to messages from the subscription moment onwards.
 
-Authenticating with Ethereum also automatically creates an associated Streamr user, even if it doesn't already exist. Under the hood, the client will cryptographically sign a challenge to authenticate you as a Streamr user:
+One can either fetch the historical sent messages with the `resend` method:
```
```suggestion
One can either fetch historical messages with the `resend` method:
```

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
 
-Authenticating with an Ethereum private key contained in an Ethereum (web3) provider:
-
+Resend from a specific message reference up to the newest message:
```
This example use a message reference (it is using timestamp)

bearni95 replied:
Sorry, not following. What did you mean here?

teogeb replied:
In the example there is a timestamp property in from block. A ResendRef can be a MessageRef or a timestamp-based reference:
https://github.com/streamr-dev/network-monorepo/blob/aa7b7a0c3db3cf972355629390f3798dcb2d0c4e/packages/client/src/subscribe/Resends.ts#L35

The is good to have the timestamp-based reference in the example. Just update the wording in line 245.

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+)
+```
 
+Or fetch them and subscribe to new messages in the same call via a `subscribe` call:
````
```suggestion
Or fetch historical message and subscribe to real-time messages in the same call via a `subscribe` call:
```

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+    {
+        last: 10,
+    }, 
+    messageCallback
```
consistency for the callback name: maybe use onMessage?

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-
-Authenticating with a pre-existing session token (used internally by the Streamr app):
-
+Resend a limited range of messages:
```
```suggestion
Resend a range of messages:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-const bearerToken = await client.session.getSessionToken()
+const sub = await streamr.subscribe(options)
+sub.onResent(() => {
+    console.log('All caught up and received all requested historical messages! Now switching to real time!')
```
```suggestion
    console.log('Received all requested historical messages! Now switching to real time!')
```

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-    auth: {
-        privateKey: 'your-private-key'
+// Fetches the last 10 messages and subscribes to the stream
+const sub1 = await streamr.subscribe({
```
Should use `resend` in all examples? If we've separately mentioned that there are the two styles: `resend` and `subscribe`?

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
 
-Then for example,
+### Search Streams
+You can search for a stream which has the `term` string in it's id as follows:
```
```suggestion
You can search for streams by specifying a search term:
```

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
-    )}
+const streams = await streamr.searchStreams('foo')
 ```
+You can query for the streams using an optional second parameter to fine-tune your search. A permission query searches over stream permissions. You can either query by direct permissions (which are explicitly granted to a user), or by all permissions (including public permissions, which apply to all users).
````
```suggestion
Alternatively or additionally to the search term, you can search for streams based on permissions.
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-Note, session tokens expire after four hours and may need to be refreshed.
-
-## Connecting
+To get all streams where a user has some direct permission. The `user` option can be omitted. In that case, it defaults to the authenticated user:
```
```suggestion
To get all streams for which a user has any direct permission:
```

teogeb replied:
It doesn't default to the authenticated user. The cli-tools command does, when we define a `--user` flag.

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+const streams = await streamr.searchStreams('foo', {
+    user: '0x12345...'
+})
+// Or, to query for all the streams accessible by the user
```
I would divide the code snippet into two and precede the 2nd snippet with
```
To get all streams for which a user has any permission (direct or public):
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-messages were recently published. This behaviour can be disabled using the `autoConnect` & `autoDisconnect` options when creating a `new
-StreamrClient`. Explicit calls to either `connect()` or `disconnect()` will disable all `autoConnect` & `autoDisconnect` functionality, but they can
-be re-enabled by calling `enableAutoConnect()` or `enableAutoDisconnect()`.
+It is also possible to filter by specific permissions by using `allOf` and `anyOf` flags. Please prefer `allOf` to `anyOf` when possible as it has better query performance.
```
```suggestion
It is also possible to filter by specific permissions by using `allOf` and `anyOf` flags. Flag `allOf` should be preferred over `anyOf` when possible due to better query performance.
```

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+It is also possible to filter by specific permissions by using `allOf` and `anyOf` flags. Please prefer `allOf` to `anyOf` when possible as it has better query performance.
 
-Calls that need a connection, such as `publish` or `subscribe` will fail with an error if you are disconnected and autoConnect is disabled.
+If you want to find the streams you can exclusively subscribe to:
```
For the both "you can publish to" examples we could leave out the search terms as it is more realistic use case and more in line what we describe in the sentence.

bearni95 replied:
What would that look like? Would `...` just do it, or does it become confusing?:
```js
const streams = await streamr.searchStreams(
    ...
    , {
      user: '0x12345...',
      allOf: [StreamPermission.SUBSCRIBE],
    }
 )
```

teogeb replied:
I mean, use `undefined` as the search term. As we describe the use case as "streams you can subscribe to", not e.g. "streams you can subscribe to, and also containing term foobar":
https://github.com/streamr-dev/network-monorepo/blob/0c7e55c2b4417b3f049fc09241b28f9832ffef55/packages/client/src/StreamRegistry.ts#L259

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+It is also possible to filter by specific permissions by using `allOf` and `anyOf` flags. Please prefer `allOf` to `anyOf` when possible as it has better query performance.
 
-Calls that need a connection, such as `publish` or `subscribe` will fail with an error if you are disconnected and autoConnect is disabled.
+If you want to find the streams you can exclusively subscribe to:
```
```suggestion
If you want to find the streams you can subscribe to:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+const streams = await streamr.searchStreams('foo', {
+    user: '0x12345...',
+    allOf: [StreamPermission.SUBSCRIBE],
+})
```
```suggestion
const streams = await streamr.searchStreams('foo', {
    user: '0x12345...',
    allOf: [StreamPermission.SUBSCRIBE],
    allowPublic: true
})
```

teogeb commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+    allOf: [StreamPermission.SUBSCRIBE],
+})
+```
+If you want to find any streams you can publish to, regardless of the other permissions assigned:
````
Maybe we don't need all examples of anyOf/allOf?

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+const streams = await streamr.searchStreams('foo', {
+    user: '0x12345...',
+    anyOf: [StreamPermission.PUBLISH],
+})
```
```suggestion
If you want to find the streams you can publish to:
```js
const streams = await streamr.searchStreams('foo', {
    user: '0x12345...',
    allOf: [StreamPermission.PUBLISH],
    allowPublic: true
})
```

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+    anyOf: [StreamPermission.PUBLISH],
+})
+```
+The `allOf` method will return streams which permissions exactly match the provided array:
````
```suggestion
To find the streams for which a user has both publish and subscribe permission:
```

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+```js 
+const streams = await streamr.searchStreams('foo', {
+    allOf: [StreamPermission.SUBSCRIBE, StreamPermission.PUBLISH]
+})
````
```suggestion
const streams = await streamr.searchStreams('foo', {
    user: '0x12345...',
    allOf: [StreamPermission.SUBSCRIBE, StreamPermission.PUBLISH],
    allowPublic: true
})
```

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+        - Stream Permissions
+        - Deleting a stream
+    - Storage Options
+    - Data Unions
```
Check the order of importance

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-| enableAutoConnect(enable = true)    | Enables autoConnect if it wasn't already enabled. Does not connect immediately. Use `enableAutoConnect(false)` to disable autoConnect.                                                              |
-| enableAutoDisconnect(enable = true) | Enables autoDisconnect if it wasn't already enabled. Does not disconnect immediately. Use `enableAutoConnect(false)` to disable autoDisconnect.                                                     |
+#### Fetching existent streams
+Getting an existent stream is pretty straight-forward
```
```suggestion
Getting an existing stream:
```

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-| disconnect()                        | Safely disconnects if not already disconnected, clearing all subscriptions. Returns a Promise.  Resolves immediately if already disconnected. Only rejects if an error occurs during disconnection. |
-| enableAutoConnect(enable = true)    | Enables autoConnect if it wasn't already enabled. Does not connect immediately. Use `enableAutoConnect(false)` to disable autoConnect.                                                              |
-| enableAutoDisconnect(enable = true) | Enables autoDisconnect if it wasn't already enabled. Does not disconnect immediately. Use `enableAutoConnect(false)` to disable autoDisconnect.                                                     |
+#### Fetching existent streams
```
Consider other verb?

harbu replied:
Maybe a redundant section title, could this just be directly under "Interacting with the `Stream` object"?

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+```
+___
+### Interacting with the `Stream` object
 
````
"The `Stream` type provides a convenient way to interact with a stream without having to repeatedly pass Stream IDs."

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+const stream = await streamr.getStream(streamId)
+```
 
+The method getOrCreateStream allows for a seamless creation/fetching process:
````
```suggestion
The method getOrCreateStream fetches the stream if it exists, and if not, creates it:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
 
-await client.connect()
+#### Updating a stream
+Updates the description locally set for the stream
```
```suggestion
To update the description of a stream:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-| payload       | A JS object containing the message payload itself                                                                                                                                                                                |
-| streamMessage | The whole [StreamMessage](https://github.com/streamr-dev/streamr-client-protocol-js/blob/master/src/protocol/message_layer/StreamMessage.js) object containing various metadata, for example `streamMessage.getTimestamp()` etc. |
 
+To grant permissions for users:
```
```suggestion
To grant a permission for a user:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+    permissions: [StreamPermission.PUBLISH],
 })
 
+// And for public streams:
```
```suggestion
// Or to grant a public permission
```

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+    permissions: [StreamPermission.SUBSCRIBE]
+})
 ```
+And to revoke them:
````
```suggestion
To revoke a permission from a user:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+})
 
-### Subscription Options
+// Or revoke public permissions:
```
```suggestion
// Or revoke public permission:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
-| partition | Partition number to subscribe to. Defaults to partition 0.                         |
-| resend    | Object defining the resend options. Below are examples of its contents.            |
-| groupKeys | Object defining the group key as a hex string for each publisher id of the stream. |
+There is also method `streamr.setPermissions`. You can use it to set an exact set of permissions for one or more streams. Note that if there are existing permissions for the same users in a stream, the previous permissions are overwritten:
```
```suggestion
The method `streamr.setPermissions` can be used to set an exact set of permissions for one or more streams. Note that if there are existing permissions for the same users in a stream, the previous permissions are overwritten:
```

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
 ```
 
-### Storage
+All streams permissions can be queried by calling `stream.getPermissions()`:
````
```suggestion
The full list of permissions for a stream can be queried by calling `stream.getPermissions()`:
```

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+```js
+const permissions = await stream.getPermissions()
+```
+The returned permissions are an array containing an item for each user, and one for public permissions:
````
```suggestion
The returned value is an array of permissions containing an item for each user, and possibly one for public permissions:
```

harbu commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
 
-`await stream.addToStorageNode(StorageNode.STREAMR_GERMANY)`
+#### Deleting a stream
+Deletes the stream from the on-chain registry:
```
```suggestion
To delete a stream:
```

teogeb commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+#### Deleting a stream
+Deletes the stream from the on-chain registry:
+```js
+// Requires gas
````
Consistent comment about gas?

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+### Storage Options
 
-### Stream object
+You can enable data storage on your streams to retain historical data in one or more geographic locations of your choice and access it later via `resend`. By default storage is not enabled on streams. You can enable it with:
```
There are no multiple geographic locations at the moment (just the storage node in Germany)

harbu commented on packages/client/README.md:
````diff
@@ -114,257 +205,280 @@ const msg = {
+```js
+const { StreamrClient, STREAMR_STORAGE_NODE_GERMANY } = require('streamr-client')
+...
+// assign a stream to storage
````
```suggestion
// assign a stream to a storage node
```

teogeb commented on packages/client/README.md:
```diff
@@ -10,97 +10,188 @@
+![GitHub stars](https://img.shields.io/github/stars/streamr-dev/network-monorepo?style=social)
 [![Discord Chat](https://img.shields.io/discord/801574432350928907.svg?label=Discord&logo=Discord&colorB=7289da)](https://discord.gg/FVtAph9cvz)
 
 This library allows you to easily interact with the [Streamr Network](https://streamr.network) from JavaScript-based environments, such as browsers and [node.js](https://nodejs.org). The library wraps a Streamr light node for publishing and subscribing to data, as well as contains convenience functions for creating and managing streams.
```
We should consider if we refer to message (preferably) or data, and not use both words in this document?

bearni95 replied:
What does this make reference to?

teogeb replied:
I mean that in this document we should use only one word to reference to the input type of a subscription (and also output type of a publish, etc). Usually we use word "message", but sometimes e.g. word "data", like here. 

I suggest that we'd update this sentence to be:
"...publishing and subscribing messages..."

And also all other occurrences in the document, e.g.
" ..."By default `subscribe` will not resend historical messages. It only subscribes to real time messages..." (line 222)

(If some sentence are bit boring after that change it could make sense to shorten those. Not a high priority thing.)

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +205,280 @@ const msg = {
+// assign a stream to storage
+await stream.addToStorageNode(STREAMR_STORAGE_NODE_GERMANY)
+// fetch the storage nodes for a stream
+const storageNodes = stream.getStorageNodes()
```
The list of storage nodes is the least used feature, I guess?

teogeb commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
 
-## Utility functions
+### Utility functions
+The static function `StreamrClient.generateEthereumAccount()` generates a new Ethereum private key and returns an object with fields `address` and `privateKey`. Note that this private key can be used to authenticate to the Streamr API by passing it in the authentication options, as described earlier in this document.
```
This could be a bit shorter explanation as we've already introduced this earlier?

teogeb replied:
I'd remove lines 674-679. This chapter is about the utility functions, we may assume that users know if they need them? (We've linked to this section)

bearni95 replied:
Sorry, which lines are you suggesting to remove? The utility method text has changed since this comment:
https://github.com/streamr-dev/network-monorepo/blob/98f9488b48d71a08fc67b2001f2ac453ab418e77/packages/client/README.md#utility-functions

teogeb replied:
These:
```
Note that this private key can be used to authenticate to the Streamr API by passing it in the authentication options, as described earlier in this document:
...
const streamr = new StreamrClient({
    auth: { privateKey }
})
```

harbu commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
 
-The client and the subscriptions can fire events as detailed below.
-You can bind to them using `on`.
+By default the client will automatically connect and disconnect as needed, ideally you should not need to manage connection state explicitly.
```
```suggestion
By default the client will automatically connect and disconnect as needed, you should not need to manage connection state manually.
```

harbu commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
 
-To grant and revoke permissions:
 
+### Disable message ordering
```
```suggestion
### Disable message ordering and gap filling
```

harbu commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
-To grant and revoke permissions:
 
+### Disable message ordering
+If your use-case doesn't require message order to be enforced or if you want it to be tolerant to out-of-sync messages you can turn off the message ordering upon client creation:
```
```suggestion
If your use case tolerates missing messages and message arriving out-of-order, you can turn off message ordering and gap filling when creating a instance of the client:
```

harbu commented on packages/client/README.md:
````diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
+    gapFill: false
 })
 ```
+Both of these flags should be disabled in tandem for message ordering to be properly turned off.
````
```suggestion
Both of these flags should be disabled in tandem for message ordering and gap filling to be properly turned off.
```

teogeb commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
-| removeListener(eventName, function) | Unbinds the `function` from events called `eventName`                                    |
 
-### Events on the StreamrClient instance
+Specifically, it will automatically connect when you publish or subscribe, and automatically disconnect once all subscriptions are removed and no messages were recently published. This behaviour can be disabled using the `autoConnect` & `autoDisconnect` options when creating a `new StreamrClient`. Explicit calls to either `connect()` or `disconnect()` will disable all `autoConnect` & `autoDisconn ...
```
Is this description still valid? To me it looks like `disconnect` is an alias to `destroy`, and therefore I'd expect we shouldn't do anything after destroying the instance:
https://github.com/streamr-dev/network-monorepo/blob/557a7ecc92779e1a40d6d6d5e7485bc4a40516d9/packages/client/src/StreamrClient.ts#L146

Also, as the `disconnect` is a deprecated method, we should use `destroy` instead

bearni95 replied:
should the references to `disconnect` be removed and the text reworked into "here's how to auto-connect" or should the section be removed?

teogeb replied:
No. Two things:
- `Disconnect` is a deprecated method, and the documentation should not reference to any deprecated methods. A replacement for the method is `destroy()`. (see the link in the previous message)
- To me it sounds that destroy() is not a direct replacement for disconnect() because the word "disconnect" would tell me that I can reconnect if I want. But a word "destroy" indicated that after destroying a client, we should't interact with that client instance anymore.

If the assumption in my second bullet-point is valid, some sentences in the above section needs to be corrected. Maybe easiest to test this out in practice?

bearni95 replied:
Removed the references to `disconnect` in the text and snippets, not sure this is the right place to mention `destroy` since it is unrelated to the connect/disconnect functionality.

Here's the latest text to look into, am assuming it will require a few more corrections:
https://github.com/streamr-dev/network-monorepo/blob/1119d4159e55bcbacc59d119cac3c23aeb08b443/packages/client/README.md#manual-connection-management

teogeb replied:
To me it looks like `autoConnect` and `autoDisconnect` don't do anything. Have you tried to execute the example? Or is there an example in the examples repo about this?

teogeb replied:
Made a PR: https://github.com/streamr-dev/network-monorepo/pull/471

harbu commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
-    permission: StreamPermission.PUBLISH,
-    user: await client.getAddress(),
-    allowPublic: true
+Stream id as a string:
```
```suggestion
// Stream id as a string:
```

harbu commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
+#### Creating partitioned streams
 
-By default, streams only have 1 partition when they are created. The partition count can be set to any positive number (1-100 is reasonable). An example of creating a partitioned stream using the JS client:
+By default, streams only have 1 partition when they are created. The partition count can be set to any positive number (max 100). An example of creating a partitioned stream using the JS client:
```
```suggestion
By default, streams only have 1 partition when they are created. The partition count can be set to any number between 1 and 100. An example of creating a partitioned stream:
```

harbu commented on packages/client/README.md:
```diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
-    name: 'My partitioned stream',
+// Requires gas
+const stream = await streamr.createStream({
+    id: `/partitioned-stream`,
```
```suggestion
    id: `/foo/bar`,
```

teogeb commented on packages/client/README.md:
````diff
@@ -550,137 +664,100 @@ const dataUnion = await client.deployDataUnion({
 ```js
-const stream = await client.createStream({
-    name: 'My partitioned stream',
+// Requires gas
````
Consistent comment about gas. Also why we say about the gas here, but not e.g. when we modify permissions?

harbu commented on packages/client/README.md:
````diff
@@ -689,20 +766,35 @@ The library allows the user to choose a _partition key_, which simplifies publis
+```js 
+await streamr.publish(streamId, msg, Date.now(), 4)
+// or, equivalently, using the stream object
+await stream.publish(msg, Date.now(), 4)
````
I'd remove this part. Let's not advertise this way of publishing to partition. The example after this is preferred.

teogeb commented on packages/client/README.md:
```diff
@@ -689,20 +766,35 @@ The library allows the user to choose a _partition key_, which simplifies publis
+await streamr.publish({
+    id: `${address}/foo/bar`,
+    partition: 4
+}, msg, Date.now())
```
Date.now() is redundant

harbu commented on packages/client/README.md:
````diff
@@ -689,20 +766,35 @@ The library allows the user to choose a _partition key_, which simplifies publis
+// or, equivalently, using the stream object
+await stream.publish(msg, Date.now(), 4)
+```
+Alternatively, you can specify the partition number as part of the stream id:
````
```suggestion
Alternatively, you can specify the partition number explicitly:
```

harbu commented on packages/client/README.md:
```diff
@@ -689,20 +766,35 @@ The library allows the user to choose a _partition key_, which simplifies publis
+#### Subscribing to partitioned streams
 
-By default, the JS client subscribes to the first partition (partition `0`) in a stream. The partition number can be explicitly given in the subscribe call:
+By default, the JS client subscribes to the first partition (partition `0`) in a stream. This behavior will change in the future so that it will subscribe to all partitions by default.
```
```suggestion
By default, the client subscribes to the first partition (partition `0`) of a stream. Be aware: this behavior will change in the future so that it will subscribe to _all_ partitions by default.
```

harbu reviewed (changes requested):
Getting there, see comments

bearni95 committed:
docs: corrections from comments on styles & text

bearni95 committed:
docs: see below:
- reworked index
- index with local links
- updated "requires gas" comments consistently
- re-sorted advanced usage

bearni95 committed:
docs: updates from comments (see below)

- normalized casing in titles
- added "and gap filling" to the "Disable message ordering" section
- text modification from comments

bearni95 committed:
docs: updated searchStream's term to undefined on trivial snippets

bearni95 committed:
docs: changed ref to messageRef to messageTs in resend text

teogeb commented on packages/client/README.md:
````diff
@@ -248,7 +248,7 @@ const sub1 = await streamr.subscribe({
 ```
 In order to fetch historical messages the stream needs to have [storage enabled](#storage).
 
-Resend from a specific message reference up to the newest message:
````
Just "timestamp". A message timestamp is not a concept

teogeb commented on packages/client/README.md:
```diff
@@ -10,100 +10,189 @@
+        - [Stream permissions](#stream-permissions)
+        - [Deleting a stream](#deleting-a-stream)
+    - [Storage options](#storage-options)
+    - [Data Unions](#data-unions)
```
The DU sub-sections could be in prioritized order? (https://github.com/streamr-dev/network-monorepo/pull/353#discussion_r811892505)

teogeb commented on packages/client/README.md:
```diff
@@ -10,100 +10,189 @@
+        - [Updating a stream](#updating-a-stream)
+        - [Stream permissions](#stream-permissions)
+        - [Deleting a stream](#deleting-a-stream)
+    - [Storage options](#storage-options)
```
Could be simply just "Enabling storage"?

bearni95 replied:
I've changed it into "Message storage for historical data", which seems a bit more wordy, in line with previous titles. Which one should we go with?

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +203,284 @@ const msg = {
-## Data Unions
+### Data Unions
 
 This library provides functions for working with Data Unions. Please see the [TypeScript generated function documentation](https://streamr-dev.github.io/streamr-client-javascript/classes/dataunion_dataunion.dataunion.html) for information on each Data Union endpoint.
```
The link is not valid (and it is a API doc, not "TypeScript generated function documentation")

teogeb commented on packages/client/README.md:
````diff
@@ -689,20 +725,33 @@ The library allows the user to choose a _partition key_, which simplifies publis
+// or, equivalently, using the stream object
 await stream.publish(msg, Date.now(), msg.vehicleId)
 ```
+You can also specify the partition number as the last parameter:
````
Remove this example as we don't encourage this style

teogeb commented on packages/client/README.md:
````diff
@@ -114,257 +203,284 @@ const msg = {
+const stream = await streamr.getStream(streamId)
+```
 
+The method getOrCreateStream fetches the stream if it exists, and if not, creates it:
````
To be consistent with the method name: replace `fetches` -> `gets`

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +203,284 @@ const msg = {
-    },
-    autoConnect: false,
-    autoDisconnect: false,
+// May require MATIC tokens (Polygon blockchain gas token) upon stream creation
```
```suggestion
// May require MATIC tokens (Polygon blockchain gas token)
```

teogeb commented on packages/client/README.md:
```diff
@@ -114,257 +203,284 @@ const msg = {
-| autoConnect              | true                             | If set to `true`, the client connects automatically on the first call to `subscribe()`. Otherwise an explicit call to `connect()` is required.                                                                                                                                                                                          |
-| autoDisconnect           | true                             | If set to `true`, the client automatically disconnects when the last stream is unsubscribed. Otherwise the connection is left open and can be disconnected explicitly by calling `disconnect()`.                                                                                                                                        |
-| orderMessages            | true                             | If set to `true`, the subscriber handles messages in the correct order, requests missing messages and drops duplicates. Otherwise, the subscriber processes messages as they arrive without any check.                                                                                                                                  |
+### Requesting a resend of historical events with subscriptions
```
From users perspective the concept is about the historical data, not resending it. Do not use the word resend in the title and do not talk about requesting in line 225?

bearni95 committed:
docs: re-sorted 'updating a stream' and 'stream permissions' acording to comments

bearni95 committed:
docs: renamed 'storage options' into 'message storage for historical data'

bearni95 committed:
docs: removed ref to source code, added proper markdown style for StreamrObject

bearni95 committed:
docs: replaced mentions of callback param 'payload' into consistent naming 'content'

bearni95 committed:
docs: removed example of publishing via

bearni95 committed:
docs: reworked usage of the word 'data' where it could be replaced by more specific terms, ie 'messages' or 'streams'

bearni95 committed:
docs: removed ref to absurd concept 'message timestamp'

bearni95 committed:
docs: removed example of partition number as last parameter of publish as it is not encouraged behaviour

bearni95 committed:
docs: replaced 'fetches' with 'gets'

bearni95 committed:
docs: simplified gas comment for getOrCreateStream

bearni95 committed:
docs: renamed title for requesting historical events and rephrased 'request' into 'perform'

bearni95 committed:
docs: brought back createStream example using address as id parameter

bearni95 committed:
docs: moved comment about msg.vehicleId from the top to the end of the line

bearni95 committed:
docs: swapped order of #subscribing and #creating-a-stream-and-publishing-messages-to-it

bearni95 committed:
docs: removed second-level indentation titles from index; removed subsections for setup from index

bearni95 committed:
docs: removed auth from getting started; reworked the subsection's title indexing for getting started; moved relevant text about stream creation onto it's usage entry

bearni95 committed:
docs: reworked subsection titles

bearni95 committed:
docs: proposal for manual connection management text rework

bearni95 committed:
docs: changed addresses to consistent, full strings: reworked du admin functions section

teogeb commented on packages/client/README.md:
```diff
@@ -236,7 +236,7 @@ const sub2 = await streamr.resend(
             sequenceNumber: 0, // optional
         },
-        publisher: '0x12345...', // optional
+        publisher: '0x1234567890123456789012345678901234567890', // optional
```
I'd use the shorter form, it is more readable to me. (Or am I contradictory to some previous comment that I made?)

teogeb commented on packages/client/README.md:
```diff
@@ -380,10 +380,10 @@ await streamr.setPermissions({
             permissions: [StreamPermission.EDIT]
         }, {
-            user: '0x6789a...',
+            user: '0x6789a0x1234567890123456789012345678901234567891'
```
Typo in the address

teogeb replied:
It would be super readable if a end-user can think this straightforwardly:
- there are two users
- user 1 has address `0x1111...` and user 2 has address `0x2222...`
- for user 1 we grant this permission, and for user 2 we grant that other permission

If the second bullet point is confusing -> the whole logic is more confusing. Maybe not relevant in this small example.

bearni95 committed:
docs: fudged together the du member with admin snippets

bearni95 committed:
docs: reworked dataunion snippets

bearni95 committed:
docs: removed pkey example usage from utility fns

bearni95 committed:
docs: reworked example address structure into simplified view, using '0x12345...' for single-use addresses and '0x11111', '0x22222' for series

bearni95 committed:
docs: cleanup for dataunion stuff

bearni95 committed:
docs: renamed storage section; removed 'connect/disconnect' section from advanced usage

bearni95 committed:
docs: replaced occurances of 'data' with proper wording 'messages'

bearni95 committed:
docs: replaced occurances of word 'event' with fitting word 'messages'

bearni95 committed:
docs: disclaimer about not up-to-date snippets for data unions

bearni95 committed:
docs: updated beta version

bearni95 committed:
docs: added emoji to warn notice

bearni95 committed:
docs: reworked section titles that were too long

bearni95 committed:
docs: removed trailing space and dash from index element

bearni95 committed:
docs: more trailing spaces

bearni95 committed:
Merge branch 'main' into NET-423

bearni95 committed:
docs: corrected streamMessage typo

bearni95 committed:
docs: replaced streamMessage callback param by metadata; reworked it's explanation

bearni95 committed:
docs: updated text for resends; updated local links to reflect changes in section titles

bearni95 committed:
docs: changes resend by subscribe in onResent snippet

bearni95 committed:
docs: live comments on searchStream

bearni95 committed:
docs: interacting with stream live changes

bearni95 committed:
docs: live read notes, full read
